core/src/Streamly/Internal/Data/IORef/Prim.hs

```diff
@@ -42,56 +41,58 @@ where
 #include "inline.hs"
 
 import Control.Monad.IO.Class (MonadIO(..))
-import Control.Monad.Primitive (primitive_)
-import Data.Primitive.Types (Prim, sizeOf#, readByteArray#, writeByteArray#)
-import GHC.Exts (MutableByteArray#, newByteArray#, RealWorld)
-import GHC.IO (IO(..))
+import Streamly.Internal.Data.Unboxed
+    ( ArrayContents(..)
+    , Unboxed(..)
+    , peekWith
+    , pokeWith
+    , newUnpinnedArrayContents
+    )
 
 import qualified Streamly.Internal.Data.Stream.StreamD.Type as D
 
 -- | An 'IORef' holds a single 'Prim' value.
-data IORef a = IORef (MutableByteArray# RealWorld)
+type IORef a = ArrayContents a
```

Comment:
This should be a newtype. This module would become a user facing module, ArrayContents is an internal type which should not be exposed via this.